File: backend/app/schemas/contracts.py
from enum import Enum
from typing import Literal, Optional

from pydantic import BaseModel, Field


class Detection(BaseModel):
    bbox: list[float] = Field(..., min_length=4, max_length=4)
    confidence: float
    label: str
    track_id: int = -1


class RawGesture(BaseModel):
    hand: str
    gesture: str
    index_x: Optional[float] = None
    index_y: Optional[float] = None


class GestureAction(str, Enum):
    START_APP = "START_APP"
    RESET_APP = "RESET_APP"
    MENU_NEXT = "MENU_NEXT"
    MENU_PREVIOUS = "MENU_PREVIOUS"
    MENU_CLOSE = "MENU_CLOSE"
    BACK_TO_MENU = "BACK_TO_MENU"
    SELECT_RECIPE = "SELECT_RECIPE"
    NEXT_STEP = "NEXT_STEP"
    PREVIOUS_STEP = "PREVIOUS_STEP"
    NO_ACTION = "NO_ACTION"


class CVUpdateData(BaseModel):
    frame: str
    objects: list[Detection]
    action: Optional[GestureAction] = None


class CVUpdateMessage(BaseModel):
    type: Literal["cv_update"] = "cv_update"
    data: CVUpdateData


class LLMResponseData(BaseModel):
    query: str
    response: str


class LLMResponseMessage(BaseModel):
    type: Literal["llm_response"] = "llm_response"
    data: LLMResponseData


class StatusData(BaseModel):
    message: str


class StatusMessage(BaseModel):
    type: Literal["status"] = "status"
    data: StatusData


class LLMQueryMessage(BaseModel):
    type: Literal["llm_query"] = "llm_query"
    query: str

File: frontend/src/types/contracts.ts
export interface Ingredient {
    name: string;
    amount: number;
    unit: string;
    notes?: string | null;
}

export interface InstructionStep {
    step: number;
    description: string;
    timer_minutes?: number | null;
}

export interface Recipe {
    id: string;
    title: string;
    description?: string | null;
    prep_time_minutes?: number | null;
    cook_time_minutes?: number | null;
    servings?: number | null;
    difficulty?: string | null;
    cuisine?: string | null;
    ingredients: Ingredient[];
    instructions: InstructionStep[];
}

export interface Detection {
    bbox: [number, number, number, number];
    confidence: number;
    label: string;
    track_id: number;
}

export type GestureAction =
    | 'START_APP'
    | 'RESET_APP'
    | 'MENU_NEXT'
    | 'MENU_PREVIOUS'
    | 'MENU_CLOSE'
    | 'BACK_TO_MENU'
    | 'SELECT_RECIPE'
    | 'NEXT_STEP'
    | 'PREVIOUS_STEP'
    | 'NO_ACTION';

export interface CVUpdateMessage {
    type: 'cv_update';
    data: {
        frame: string;
        objects: Detection[];
        action: GestureAction | null;
    };
}

export interface LLMResponseMessage {
    type: 'llm_response';
    data: {
        query: string;
        response: string;
    };
}

export interface StatusMessage {
    type: 'status';
    data: {
        message: string;
    };
}

export interface LLMQueryMessage {
    type: 'llm_query';
    query: string;
}

export type ServerWebSocketMessage = CVUpdateMessage | LLMResponseMessage | StatusMessage;
export type ClientWebSocketMessage = LLMQueryMessage;
